### PR TITLE
ENH: Adding 'protocol' parameter to 'to_pickle'.

### LINKED ISSUE #14488: Adding optional pickle protocol version argument to pandas.to_pickle()
Sometimes pickle files saved in Python v3.x are needed to read in Python v2.x. It would be nice if one can easily set pickle protocol version in `to_pickle()`.

It can be done with the following little changes:

In file `/pandas/io/pickle.py`:

```
def to_pickle(obj, path, protocol=pkl.HIGHEST_PROTOCOL):
    """
    Pickle (serialize) object to input file path

    Parameters
    ----------
    obj : any object
    path : string
        File path
    """
    with open(path, 'wb') as f:
        pkl.dump(obj, f, protocol=protocol)
```

In file `/pandas/core/generic.py`:

```
    def to_pickle(self, path, protocol):
        """
        Pickle (serialize) object to input file path.

        Parameters
        ----------
        path : string
            File path
        """
        from pandas.io.pickle import to_pickle
        return to_pickle(self, path, protocol)
```

### PR DESCRIPTION
**Pull request by jbschiratti**
This PR aims at adding an optional `protocol` parameter to the function `to_pickle`. 
Closes #14488. If needed, I can update the corresponding test (`pandas/tests/io/test_pickle.py`).

### CODE REVIEW

**Comment by jreback** [comment]
tests are the first thing you should write.

File: pandas/core/generic.py [hunk 1]
```diff
@@ -50,6 +50,7 @@
 from pandas.compat.numpy import function as nv
 from pandas.compat import (map, zip, lzip, lrange, string_types,
                            isidentifier, set_function_name)
+from pandas.compat import cPickle as pkl
```
Comment (jreback) [inline pandas/core/generic.py]: move to the previous line

File: pandas/core/generic.py [hunk 2]
```diff
@@ -1354,11 +1356,15 @@ def to_pickle(self, path, compression='infer'):
             File path
         compression : {'infer', 'gzip', 'bz2', 'xz', None}, default 'infer'
             a string representing the compression to use in the output file
+        protocol : int
+            Int which indicates which protocol should be used by the pickler,
+            default HIGHEST_PROTOCOL (Pickle module).
```
Comment (jreback) [inline pandas/core/generic.py]: version added tag (the compression one goes above this), your tag should read 0.21.0

File: pandas/io/pickle.py [hunk 3]
```diff
@@ -18,6 +18,9 @@ def to_pickle(obj, path, compression='infer'):
         File path
     compression : {'infer', 'gzip', 'bz2', 'xz', None}, default 'infer'
         a string representing the compression to use in the output file
+    protocol : int
```
Comment (jreback) [inline pandas/io/pickle.py]: same

**Review by jreback — CHANGES_REQUESTED** [review]
needs some tests. with valid & invalid protocol passed.

**Comment by jreback** [comment]
also add a whatsnew in enhancements for 0.21.0

**Comment by jbschiratti** [comment]
I am facing an issue when writing the tests.
I Installed a pandas-dev environment as suggested in http://pandas.pydata.org/pandas-docs/stable/contributing.html. When I try to run `pytests` on `pandas/tests/io/test_pickle.py`, I get the following error :

```
 % pytest test_pickle.py                                                                                                                                                                                             ~/pandas/pandas/tests/io
============================================================================================================ test session starts =============================================================================================================
platform linux -- Python 3.6.1, pytest-3.0.7, py-1.4.33, pluggy-0.4.0
rootdir: /cal/homes/jbschiratti/pandas, inifile: setup.cfg
plugins: cov-2.3.1
collected 42 items 

test_pickle.py ..........................................Traceback (most recent call last):
  File "/cal/homes/jbschiratti/anaconda3/envs/pandas_dev/bin/pytest", line 6, in <module>
    sys.exit(pytest.main())
  File "/cal/homes/jbschiratti/anaconda3/envs/pandas_dev/lib/python3.6/site-packages/_pytest/config.py", line 57, in main
    return config.hook.pytest_cmdline_main(config=config)
  File "/cal/homes/jbschiratti/anaconda3/envs/pandas_dev/lib/python3.6/site-packages/_pytest/vendored_packages/pluggy.py", line 745, in __call__
    return self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)
  File "/cal/homes/jbschiratti/anaconda3/envs/pandas_dev/lib/python3.6/site-packages/_pytest/vendored_packages/pluggy.py", line 339, in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
  File "/cal/homes/jbschiratti/anaconda3/envs/pandas_dev/lib/python3.6/site-packages/_pytest/vendored_packages/pluggy.py", line 334, in <lambda>
    _MultiCall(methods, kwargs, hook.spec_opts).execute()
  File "/cal/homes/jbschiratti/anaconda3/envs/pandas_dev/lib/python3.6/site-packages/_pytest/vendored_packages/pluggy.py", line 614, in execute
    res = hook_impl.function(*args)
  File "/cal/homes/jbschiratti/anaconda3/envs/pandas_dev/lib/python3.6/site-packages/_pytest/main.py", line 127, in pytest_cmdline_main
    return wrap_session(config, _main)
  File "/cal/homes/jbschiratti/anaconda3/envs/pandas_dev/lib/python3.6/site-packages/_pytest/main.py", line 122, in wrap_session
    exitstatus=session.exitstatus)
  File "/cal/homes/jbschiratti/anaconda3/envs/pandas_dev/lib/python3.6/site-packages/_pytest/vendored_packages/pluggy.py", line 745, in __call__
    return self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)
  File "/cal/homes/jbschiratti/anaconda3/envs/pandas_dev/lib/python3.6/site-packages/_pytest/vendored_packages/pluggy.py", line 339, in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
  File "/cal/homes/jbschiratti/anaconda3/envs/pandas_dev/lib/python3.6/site-packages/_pytest/vendored_packages/pluggy.py", line 334, in <lambda>
    _MultiCall(methods, kwargs, hook.spec_opts).execute()
  File "/cal/homes/jbschiratti/anaconda3/envs/pandas_dev/lib/python3.6/site-packages/_pytest/vendored_packages/pluggy.py", line 613, in execute
    return _wrapped_call(hook_impl.function(*args), self.execute)
  File "/cal/homes/jbschiratti/anaconda3/envs/pandas_dev/lib/python3.6/site-packages/_pytest/vendored_packages/pluggy.py", line 250, in _wrapped_call
    wrap_controller.send(call_outcome)
  File "/cal/homes/jbschiratti/anaconda3/envs/pandas_dev/lib/python3.6/site-packages/_pytest/terminal.py", line 358, in pytest_sessionfinish
    outcome.get_result()
  File "/cal/homes/jbschiratti/anaconda3/envs/pandas_dev/lib/python3.6/site-packages/_pytest/vendored_packages/pluggy.py", line 279, in get_result
    raise ex[1].with_traceback(ex[2])
  File "/cal/homes/jbschiratti/anaconda3/envs/pandas_dev/lib/python3.6/site-packages/_pytest/vendored_packages/pluggy.py", line 265, in __init__
    self.result = func()
  File "/cal/homes/jbschiratti/anaconda3/envs/pandas_dev/lib/python3.6/site-packages/_pytest/vendored_packages/pluggy.py", line 614, in execute
    res = hook_impl.function(*args)
  File "/cal/homes/jbschiratti/anaconda3/envs/pandas_dev/lib/python3.6/site-packages/_pytest/cacheprovider.py", line 152, in pytest_sessionfinish
    prev_failed = config.cache.get("cache/lastfailed", None) is not None
  File "/cal/homes/jbschiratti/anaconda3/envs/pandas_dev/lib/python3.6/site-packages/_pytest/cacheprovider.py", line 57, in get
    return json.load(f)
AttributeError: module 'json' has no attribute 'load'
```
Is it because it indirectly calls the wrong `json` module ?

**Comment by TomAugspurger** [comment]
Try changing to the root of the git repo and running `pandas/tests/io/test_pickle.py`

**Comment by jbschiratti** [comment]
@TomAugspurger It solved my problem. Thanks!

**Comment by jreback** [comment]
this needs some tests.

**Comment by jbschiratti** [comment]
I'm on it. I'll make a new commit, very soon, with the necessary tests.

**Comment by jbschiratti** [comment]
@jreback I added some tests for valid and invalid values of the `protocol` parameter. Please, have a look and tell me if more tests are needed or if there is a problem with the code.

**Comment by jbschiratti** [comment]
@jreback : Do you think that more tests are needed ?

File: doc/source/whatsnew/v0.21.0.txt [hunk 4]
```diff
@@ -27,6 +27,28 @@ New features
 Other Enhancements
 ^^^^^^^^^^^^^^^^^^
 
+Pickle file I/O protocol parameter
```
Comment (jreback) [inline doc/source/whatsnew/v0.21.0.txt]: I don't think we need this section. Can just add a single line with the issue refernce.

**Review by jreback — APPROVED** [review]
lgtm.

**Comment by jreback** [comment]
small linting failure.

File: doc/source/whatsnew/v0.21.0.txt (hunk 4 above)
Comment (jbschiratti) [inline doc/source/whatsnew/v0.21.0.txt]: I changed the 'whatsnew' file accordingly.

**Comment by jbschiratti** [comment]
If I understand correctly, the failure comes from `cpplint` because the code in some file violates style requirements (it is not compliant with Google's C++ style guide). How could I fix this ?

File: pandas/core/generic.py [hunk 5]
```diff
@@ -1356,9 +1357,21 @@ def to_pickle(self, path, compression='infer'):
             a string representing the compression to use in the output file
 
             .. versionadded:: 0.20.0
+        protocol : int
+            Int which indicates which protocol should be used by the pickler,
+            default HIGHEST_PROTOCOL (Pickle module). The possible values for
+            this parameter depend on the version of Python. For Python 2.x,
+            possible values are 0, 1, 2. For Python>=3.0, 3 is a valid value.
+            For Python >= 3.4, 4 is a valid value.A negative value for the
+            protocol parameter is equivalent to setting its value to
```
Comment (jreback) [inline pandas/core/generic.py]: can you add a link to the pickle docs where this is defined.

File: doc/source/whatsnew/v0.21.0.txt [hunk 6]
```diff
@@ -27,7 +27,7 @@ New features
 Other Enhancements
 ^^^^^^^^^^^^^^^^^^
 
-
+- Added protocol parameter to :func:`to_pickle`. The 'protocol' parameter defaults to HIGHEST_PROTOCOL. (:issue:`16252`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.21.0.txt]: :func:`to_pickle` has gained a protocol parameter.....

File: pandas/io/pickle.py [hunk 7]
```diff
@@ -20,13 +20,27 @@ def to_pickle(obj, path, compression='infer'):
         a string representing the compression to use in the output file
 
         .. versionadded:: 0.20.0
+    protocol : int
+        Int which indicates which protocol should be used by the pickler,
+        default HIGHEST_PROTOCOL (Pickle module). The possible values for
+        this parameter depend on the version of Python. For Python 2.x,
+        possible values are 0, 1, 2. For Python>=3.0, 3 is a valid value.
```
Comment (jreback) [inline pandas/io/pickle.py]: same

**Review by jreback — COMMENTED** [review]
lgtm. minor comments. ping on green.

File: pandas/core/generic.py (hunk 5 above)
Comment (jbschiratti) [inline pandas/core/generic.py]: done, I added the url.

**Comment by jbschiratti** [comment]
With my previous commit, all tests were successful. After modifying docstrings and the whatsnew file, `test_clipboard` is failing, according to Travis. Is there something wrong with my recent changes ?

File: doc/source/whatsnew/v0.21.0.txt [hunk 8]
```diff
@@ -27,7 +27,8 @@ New features
 Other Enhancements
 ^^^^^^^^^^^^^^^^^^
 
-
+- :func:`to_pickle` has gained a protocol parameter (:issue:`16252`). By default,
+this parameter is set to HIGHEST_PROTOCOL (see https://docs.python.org/3/library/pickle.html, 12.1.2).
```
Comment (jreback) [inline doc/source/whatsnew/v0.21.0.txt]: does this render? e.g. we normally use `HIGHEST_PROTOCOL <url....>`__

**Comment by jreback** [comment]
I restarted that CI job. sometimes that tests fails :<

minor comment on docs. just ping when pushed.

**Comment by jbschiratti** [comment]
@jreback : I pushed the changes.

**Comment by jreback** [comment]
thanks!

**Comment by jreback** [comment]
keem em coming!

**Pull request closed without merging**